### ISSUE TITLE
[Bug] Blanket bf16 cast in _weight_to_mcore_format silently corrupts integer tensors — DeepSeek-V4 tid2eid 383→384 crashes NPU RL rollout, silently misroutes MoE on all platforms

### ISSUE THREAD

**Issue by whthust** [ISEEKYAN/mbridge#159]
### Summary

- **Problem**: when RL-training DeepSeek-V4(-Pro) with veRL + Megatron backend on Ascend NPU, the vLLM rollout engine **crashes deterministically** on the first generation after every weight sync: `MoeDistributeDispatchV2` kernel fault (`507035`, *"The address for the scalar to access the internal buffer of AICore is out of bounds"*). Pure inference with the **same checkpoint, same vllm-ascend, same EP16 topology works fine** — the crash only appears in the RL (weight-sync) flow.
- **Root cause**: `mbridge/core/bridge.py::_weight_to_mcore_format` unconditionally casts **every** incoming HF tensor to `self.dtype` (bf16) unless the name contains `expert_bias` (main @ `a61943d`, [L1476-1485](https://github.com/ISEEKYAN/mbridge/blob/a61943d7fcb34a190471cfeb0a0eb8bbda621ddf/mbridge/core/bridge.py#L1476-L1485); same code in `641a5a0` @ ~L1146). DeepSeek-V4's hash-routing layers carry an **integer lookup table** `tid2eid` (int64, `[vocab_size=129280, 6]`, expert ids in `[0, 384)`). The int64→bf16→int64 round-trip rounds **383 → 384** (bf16 has an 8-bit significand; integers in `[256,512)` quantize to even numbers, and 383 is the only in-range value that rounds *out* of range). The corrupted table is then faithfully exported and broadcast to vLLM at every weight sync; neither vllm-ascend's hash gating kernel nor the CANN dispatch kernel validates expert ids → `toRankId = 384/24 = 16` (valid ranks: 0..15) → out-of-bounds addressing → AICore fault.
- **Effect**: on NPU, deterministic engine crash on the first rollout batch (~1.4% of vocab rows per hash layer contain a 383, so any sizeable batch hits one). **On GPU the same corruption does NOT crash** — the CUDA hash-routing kernel never matches the invalid id and silently leaves garbage/stale values in that slot; additionally **all 64 odd expert ids in 257..383 are silently rounded to even neighbors**, i.e. MoE routing is silently wrong (~1/6 of top-k slots, ~2/3 of tokens in the 3 hash layers) on **every** platform, including pure-GPU training.
- **Verified fix**: add a non-float guard to the cast (`w.is_floating_point()`). Reproductions below show the crash inputs are produced byte-for-byte by the bug and eliminated by the fix.

### Root cause (code chain)

```notranslate
HF checkpoint: tid2eid int64 [129280, 6], values in [0, 383]   (verified clean, official weights)
  → DeepSeek-V4 bridge mapping (verl-ascend-recipe @ 6823af1, patch/mbridge.patch L171-172):
        "mlp.router.tid2eid" ↔ "layers.{n}.ffn.gate.tid2eid"
  → mbridge _weight_to_mcore_format  (mbridge/core/bridge.py, main@a61943d L1476-1485)

        if (hasattr(self, "dtype") and self.dtype is not None
                and "expert_bias" not in mcore_weights_name):
            hf_weights = [w.to(self.dtype) if w.dtype != self.dtype else w
                          for w in hf_weights]          # ← int64 table cast to bf16 and back

        bf16 round-trip on integers in [256, 512):      # 8-bit significand, ULP = 2
            257→256, 259→260, ..., 381→380              # 64 odd ids silently rounded (in-range)
            383→384                                     # ← the ONLY value that escapes [0,384)

  → Megatron-side tid2eid nn.Parameter now contains 384s (permanently baked in)
  → every RL weight sync: export_weights (no cast) → bucketed HCCL broadcast (byte-exact)
    → vLLM load_weights (param.copy_, no cast)          → vLLM table corrupted
  → vllm-ascend hash routing: moe_gating_top_k_hash copies the row verbatim (no range check)
  → torch_npu.npu_moe_distribute_dispatch_v2 / MoeDistributeDispatchV2 kernel:
        toRankId = expertId / moeExpertNumPerRank = 384 / 24 = 16   (valid ranks: 0..15)
    → reads past the 16-entry HCCL window table → forged GM address → AICore fault 507035
```

Why the damage is *exactly* 384 and nothing else: among legal expert ids `[0,383]`, bf16 rounding leaves `0..256` and all even values intact, silently shifts 64 odd values within range, and pushes **only 383 out of range**. This matches the crash dumps precisely (see Reproduction).

### Environment

- mbridge: `641a5a0` (pinned by veRL); **defect also present on main @ `a61943d` (L1476-1485)**
- veRL + Megatron backend (vanilla mbridge path), vLLM 0.23.0 + vllm-ascend 0.23.0
- Model: DeepSeek-V4-Pro (`n_routed_experts=384`, `num_experts_per_tok=6`, `num_hash_layers=3`, bf16)
- Hardware: 16× Ascend 910_93 (single node), rollout TP8×DP2, MoE EP16 via MC2

### Reproduction

1. **Minimal numeric repro** (any torch, CPU is enough):

   ```python
   import torch
   t = torch.arange(384, dtype=torch.int64)
   c = t.to(torch.bfloat16).to(torch.int64)
   print(c[c >= 384])              # tensor([384])  ← 383 rounds out of range
   print((t != c).sum().item())    # 64             ← odd ids in 257..383 silently shifted
   ```

2. **Offline replay against the real checkpoint (decisive)**: take the clean `tid2eid` tables from the DeepSeek-V4 checkpoint (verified: int64, max value 383, zero out-of-range entries), apply one bf16 round-trip, then search for the two `expert_ids` rows captured in the **device crash dumps** of two independent production crashes (warmup at T0, first real decode at T0+62s):

   ```notranslate
   crash dump #1 input row: [352, 384, 200, 320, 147, 356]  → matches casted table row, token id 15803
   crash dump #7 input row: [148, 130, 26, 117, 48, 384]    → matches casted table row, token id 58813
   (both in layer 0; poisoned rows per layer: 1858 / 1750 / 1722 ≈ 1.4% of vocab)
   ```

   The clean table + one bf16 round-trip reproduces both crash-time kernel inputs **byte-for-byte**; without the cast they match nothing.

3. **In-situ**: with the bug, the RL task crashes on the first generation after weight sync, every time; device exception dumps (`MoeDistributeDispatchV2`, tiling `moeExpertNum=384, k=6, ep=16`) available on request.

### Exclusion evidence (all measured)

- Checkpoint on disk is clean: all 3 `tid2eid` tables int64, `max=383`, zero values ≥384 — and pure inference (vLLM loads the same checkpoint directly, never touching mbridge) works fine ⇒ corruption is introduced at runtime, not at rest.
- Weight-sync transport is byte-exact (per-tensor dtype metadata, no cast in `export_weights` / bucketed HCCL transfer / vLLM `load_weights`) ⇒ corruption happens *before* sync, at mbridge load time.
- Official upstream DeepSeek-V4-Pro tables verified clean (all values ∈ [0,383]).
- The failure is not HCCL/hardware: the faulting task is a user kernel (`stream not from HCCL`), single-rank, input-dependent, 100% reproducible from the table content.

### Proposed fix

```python
# mbridge/core/bridge.py — _weight_to_mcore_format (main @ a61943d, L1478-1485)
if (
    hasattr(self, "dtype")
    and self.dtype is not None
    and "expert_bias" not in mcore_weights_name
):
    hf_weights = [
        w.to(self.dtype) if (w.dtype != self.dtype and w.is_floating_point()) else w
        for w in hf_weights
    ]
```

i.e. only floating-point tensors participate in the dtype conversion. This preserves the original intent (FP32→BF16 for float weights, introduced by #7) and protects any integer tensor (routing tables, index maps, future exotic dtypes), whereas name-based exemptions (today's `expert_bias`) fix one tensor at a time — `tid2eid` is the second casualty of the same pattern.

A unit test suggestion: `_weight_to_mcore_format("<name>", [torch.tensor([[383]], dtype=torch.int64)])` must return 383 unchanged.

### Validation

- Minimal repro: before fix `383→384`; with the guard the int64 tensor passes through unchanged. ✔
- Offline replay: clean table + guard → no out-of-range ids; crash fingerprint rows no longer produced. ✔
- End-to-end rerun of the RL task with the patched bridge: pending on our side (will update this issue once executed).

### Compatibility notes

- The guard is a strict no-op for all-float models (every existing bridge mapping ships only float tensors — we audited all 17 upstream model bridges), so there is no regression risk for current users.
- Downstream consumers *should* additionally validate: vllm-ascend could assert `tid2eid < n_routed_experts` at load; CANN's `MoeDistributeDispatchV2` could range-check expert ids. Defense in depth would have turned this silent corruption into a loud, early error. (Will file/link separately.)
- Note for GPU users: this bug does not crash on CUDA (the hash kernel skips unmatched ids) — it **silently degrades MoE routing**. Anyone who trained DeepSeek-V4-family models through mbridge should re-verify their `tid2eid` tables (`(t >= 384).any()` and odd-value parity), including already-saved checkpoints.

### Attachments (available on request)

- Two device exception dumps (`exception_info.*`) with the faulting kernel's live inputs
- Offline replay script + per-layer poisoned-row statistics
- Full RCA write-up (timeline, code chain, verification logs)

---

*This issue was investigated and written with AI assistance; all experiments, measurements, and the fix were executed and verified on our side by human engineers.*
